### PR TITLE
Feat - onboarding pages - LESQ-1022

### PR DESCRIPTION
**Pull request by JBR90**
## Description

Music year: {owa_music_year}

- Refactors feature flag into HOC
- Add's onboarding pages   
`/onboarding `   ` /onboarding/school-selection`.   `/onboarding/role-selection`
- `/onboarding` form style and and routing from url will be in another PR - wanted to get this in so it doesn't block other PR's
- `school-selection` has the school picker form from that was previously on `onboarding` -  this should work the same as before
    
## Issue(s)

Fixes #

## How to test

1. Go to https://deploy-preview-2682--oak-web-application.netlify.thenational.academy/onboarding
You should be able to navigate to he new pages

## Checklist

- [ ] Added or updated tests where appropriate
- [ ] Manually tested across browsers / devices
- [ ] Considered impact on accessibility
- [ ] Design sign-off
- [ ] Approved by product owner
- [ ] Does this PR update a package with a breaking change

### CODE REVIEW

File: src/components/TeacherViews/Onboarding/Onboarding.view.tsx [hunk 1]
```diff
@@ -98,96 +66,48 @@ export const OnboardingView = () => {
...
-            formState.errors?.school !== undefined || !formState.isValid
+            formState.errors?.worksInSchool !== undefined || !formState.isValid
+          }
+          element={
+            formState.errors?.worksInSchool !== undefined || !formState.isValid
+              ? "button"
+              : "a"
+          }
+          href={
+            // this should be added to url.tsx
```
Comment (njblore) [inline src/components/TeacherViews/Onboarding/Onboarding.view.tsx:101]: 👁️ 👁️

File: src/components/TeacherViews/Onboarding/Onboarding.view.tsx [hunk 2]
```diff
@@ -98,96 +66,48 @@ export const OnboardingView = () => {
...
-        />
         <OakPrimaryButton
+          $mv={"space-between-m"}
           disabled={
-            formState.errors?.school !== undefined || !formState.isValid
+            formState.errors?.worksInSchool !== undefined || !formState.isValid
+          }
+          element={
+            formState.errors?.worksInSchool !== undefined || !formState.isValid
+              ? "button"
```
Comment (njblore) [inline src/components/TeacherViews/Onboarding/Onboarding.view.tsx:97]: how come it changes element type?

File: src/components/TeacherViews/Onboarding/Onboarding.view.tsx (hunk 1 above)
Comment (njblore) [reply src/components/TeacherViews/Onboarding/Onboarding.view.tsx:101]: the focus state on this when it's a link is a bit off  
Screenshot 2024-08-13 at 11 54 29

Comment (JBR90) [reply src/components/TeacherViews/Onboarding/Onboarding.view.tsx:101]: As i mentioned in the comments above, i'm going to add this in another PR once this is in, to separate the page changes. There is some weirdness around the styling in the button that needs to be resolved as well.

File: src/components/TeacherViews/Onboarding/RoleSelection.view.tsx [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    })
+    .min(1, "Select school"),
+  schoolName: z.string().optional(),
+});
+type OnboardingFormValues = z.infer<typeof onboardingFormSchema>;
+type OnboardingFormProps = OnboardingFormValues & {
+  onSubmit: (values: OnboardingFormValues) => Promise<void>;
+};
+
+export const RoleSelectionView = () => {
```
Comment (njblore) [inline src/components/TeacherViews/Onboarding/RoleSelection.view.tsx:23]: could this be just a basically empty component to save merge hassle later? 😁 🙏

File: src/components/TeacherViews/Onboarding/Onboarding.view.tsx (hunk 1 above)
Comment (njblore) [reply src/components/TeacherViews/Onboarding/Onboarding.view.tsx:101]: ah cool I missed that 👍

File: src/components/TeacherViews/Onboarding/Onboarding.view.tsx [hunk 4]
```diff
@@ -98,96 +66,48 @@ export const OnboardingView = () => {
...
-          withHomeschool={false}
-        />
         <OakPrimaryButton
+          $mv={"space-between-m"}
           disabled={
-            formState.errors?.school !== undefined || !formState.isValid
+            formState.errors?.worksInSchool !== undefined || !formState.isValid
+          }
+          element={
+            formState.errors?.worksInSchool !== undefined || !formState.isValid
```
Comment (carlmw) [inline src/components/TeacherViews/Onboarding/Onboarding.view.tsx:96]: The negation here hurts my tiny brain.

Could this be extracted and negated when passed as a prop?

```typescript
const canSubmit = formState.errors?.worksInSchool === undefined && formState.isValid;

<OakPrimaryButton
  disabled={!canSubmit}
/>
```

Also, i'm new to react-hook-form, but what is `formState.errors?.worksInSchool` doing here that `formState.isValid` isn't giving us?

File: src/components/TeacherViews/Onboarding/Onboarding.view.tsx [hunk 5]
```diff
@@ -98,96 +66,48 @@ export const OnboardingView = () => {
...
+          $mt={"space-between-m"}
+          tag="h2"
+          $font="heading-light-6"
+        >
+          Do you work in a school?
         </OakHeading>
+        <OakBox $pv={"inner-padding-xl"}>
+          <OakRadioGroup
+            onChange={(value) =>
+              setWorksInSchool(value.target.value === "yes" ? true : false)
```
Comment (carlmw) [inline src/components/TeacherViews/Onboarding/Onboarding.view.tsx:80]: ternary here is a little redundant.

```suggestion
              setWorksInSchool(value.target.value === "yes")
```

File: src/hoc/withFeatureFlag.tsx [hunk 6]
```diff
@@ -0,0 +1,41 @@
+import React, { useEffect } from "react";
```
Comment (carlmw) [inline src/hoc/withFeatureFlag.tsx:1]: I think this should be in `hocs` to match the pluralisation of `hooks`, `components` etc.

File: src/components/TeacherViews/Onboarding/Onboarding.view.tsx (hunk 2 above)
Comment (JBR90) [reply src/components/TeacherViews/Onboarding/Onboarding.view.tsx:97]: a tag has no 'disabled' state. So this switches to button so we can have that prop.

File: src/pages/onboarding/index.tsx [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import { NextPage } from "next";
+
+import OnboardingView from "@/components/TeacherViews/Onboarding/Onboarding.view";
+import withFeatureFlag from "@/hoc/withFeatureFlag";
+
+const OnboardingComponent: NextPage = () => {
+  return <OnboardingView />;
+};
+
+const OnboardingPage = withFeatureFlag(OnboardingComponent, "use-auth-owa");
```
Comment (carlmw) [inline src/pages/onboarding/index.tsx:10]: swPCV0y

**Review by carlmw — COMMENTED** [review]
Looking grand

File: src/__tests__/pages/onboarding/school-selection.test.tsx [hunk 8]
```diff
@@ -0,0 +1,16 @@
+import { screen } from "@testing-library/dom";
+
+import SchoolSelectionPage from "@/pages/onboarding/school-selection";
+import renderWithProviders from "@/__tests__/__helpers__/renderWithProviders";
+
+jest.mock("posthog-js/react", () => ({
+  useFeatureFlagEnabled: () => true,
+}));
```
[the remark this answers is not in the document]
Comment (JBR90) [reply src/__tests__/pages/onboarding/school-selection.test.tsx:8]: Will add the tech debt board

**Review by jack-skerrett-oak — CHANGES_REQUESTED** [review]
I'm getting a bit lost with which PR is introducing which functionality at the moment, but I think this is the first time that the "Do you work in a school?" question has been asked
- I don't think the question needs to be a h2, it might work better as a fieldset/legend set up as we have done for other accessible forms recently
- States seem a bit confused on the continue button, see in the screenshots
Screenshot 2024-08-13 at 13 24 08
Screenshot 2024-08-13 at 13 24 14

**Comment by JBR90** [comment]
> I'm getting a bit lost with which PR is introducing which functionality at the moment, but I think this is the first time that the "Do you work in a school?" question has been asked
> 
> * I don't think the question needs to be a h2, it might work better as a fieldset/legend set up as we have done for other accessible forms recently
> * States seem a bit confused on the continue button, see in the screenshots


This PR is just the pages, not the form yet, i've split them up to try and reduce further conflicts

> /onboarding form style and routing from url will be in another PR - wanted to get this in so it doesn't block other PR's

Will take on the  fieldset/legend comment for the next PR and aware of the style weirdness.

File: src/components/TeacherViews/Onboarding/Onboarding.view.tsx (hunk 4 above)
Comment (JBR90) [reply src/components/TeacherViews/Onboarding/Onboarding.view.tsx:96]: This PR is just the pages at this stage, this code was already in the form, will dig in on the next PR

**Comment by jack-skerrett-oak** [comment]
> > I'm getting a bit lost with which PR is introducing which functionality at the moment, but I think this is the first time that the "Do you work in a school?" question has been asked
> > 
> > * I don't think the question needs to be a h2, it might work better as a fieldset/legend set up as we have done for other accessible forms recently
> > * States seem a bit confused on the continue button, see in the screenshots
> 
> This PR is just the pages, not the form yet, i've split them up to try and reduce further conflicts
> 
> > /onboarding form style and routing from url will be in another PR - wanted to get this in so it doesn't block other PR's
> 
> Will take on the fieldset/legend comment for the next PR and aware of the style weirdness.

@JBR90 ah ok, if it is just some refactoring and pages, I will approve :)

**Review by jack-skerrett-oak — APPROVED** [review]

File: src/hoc/withFeatureFlag.tsx [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  OakMaxWidth,
+  OakThemeProvider,
+  oakDefaultTheme,
+} from "@oaknational/oak-components";
+
+const withFeatureFlag = (
+  WrappedComponent: React.ComponentType,
+  featureFlag: string,
+) => {
+  const WithFeatureFlag: React.FC = (props) => {
```
[the remark this answers is not in the document]
Comment (JBR90) [reply src/hoc/withFeatureFlag.tsx]: Very nice!

**Review by njblore — APPROVED** [review]
🎉

**Review by carlmw — APPROVED** [review]
:shipit:

**Pull request merged**